airflow/dag/serialization.py

```diff
@@ -0,0 +1,316 @@
+        """Returns a serialized DAG.
+
+        :param dag: a DAG.
+        :type dag: airflow.models.DAG
+        """
+        if dag.dag_id in visited_dags:
+            return {Encoding.TYPE: DagTypes.DAG, Encoding.VAR: str(dag.dag_id)}
+
+        new_dag = {Encoding.TYPE: DagTypes.DAG}
+        visited_dags[dag.dag_id] = new_dag
+        new_dag[Encoding.VAR] = cls.serialize_object(
+            dag, visited_dags, included_fields=cls._dag_included_fields)
+        return new_dag
+
+    @classmethod
+    def deserialize_dag(cls, encoded_dag, visited_dags):
+        """Returns a deserialized DAG.
+
+        :param encoded_dag: a JSON dict of serialized DAG.
+        :type encoded_dag: dict
+        """
+        dag = SerializedDAG(dag_id=encoded_dag['_dag_id'])
+        visited_dags[dag.dag_id] = dag
+        cls.deserialize_object(encoded_dag, dag, cls._dag_included_fields, visited_dags)
+        return dag
+
+
+class SerializedOperator(BaseOperator, Serialization):
+    """A JSON serializable representation of operator.
+
+    All operators are casted to SerializedOperator after deserialization.
+    Class specific attributes used by UI are move to object attributes.
+    """
+    _op_included_fields = list(vars(BaseOperator(task_id='test')).keys()) + [
+        '_dag', '_task_type', 'ui_color', 'ui_fgcolor', 'template_fields']
+
+    def __init__(self, *args, **kwargs):
+        BaseOperator.__init__(*args, **kwargs)
+        # task_type is used by UI to display the correct class type, because UI only
+        # receives BaseOperator from deserialized DAGs.
+        self._task_type = 'BaseOperator'
+        # Move class attributes into object attributes.
+        self.ui_color = BaseOperator.ui_color
+        self.ui_fgcolor = BaseOperator.ui_fgcolor
+        self.template_fields = BaseOperator.template_fields
+
+    @property
+    def task_type(self):
+        # Overwrites task_type of BaseOperator to use _task_type instead of
+        # __class__.__name__.
+        return self._task_type
+
+    @task_type.setter
+    def task_type(self, task_type):
+        self._task_type = task_type
+
+    @classmethod
+    def serialize_operator(cls, op, visited_dags):
+        """Returns a serializeed operator.
+
+        :param op: an operator.
+        :type op: subclasses of airflow.models.BaseOperator
+        """
+        serialize_op = cls.serialize_object(op, visited_dags,
+            included_fields=SerializedOperator._op_included_fields)
+        # Adds a new task_type field to record the original operator class.
+        serialize_op['_task_type'] = op.__class__.__name__
+        return cls.encode(serialize_op, type_=DagTypes.OP)
+
+    @classmethod
+    def deserialize_operator(cls, encoded_op, visited_dags):
+        """Returns a deserialized operator.
+
+        :param encoded_op: a JSON dict of serialized operator.
+        :type encoded_op: dict
+        """
+        op = object.__new__(SerializedOperator)
```

Comment:
How is this different then `SerializedOperator()`

Reply:
Changed to SerializedBaseOperator(task_id=encoded_op['task_id']) as calling __init__ is necessary.

task_id is need as a __init__ argument when using SerializedOperator():
>>> BaseOperator()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/google/home/zhoufang/Desktop/github/airflow/airflow/utils/decorators.py", line 89, in wrapper
    raise AirflowException(msg)
airflow.exceptions.AirflowException: Argument ['task_id'] is required